### PR TITLE
Multi-cursor: queue reader priority rate limiting

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Use priority rate limiting for multi-cursor reader, and use reader id as priority.

**Why?**
- Ensure default reader won't get ratelimited by reader with higher ID which are for throttled queues.

**How did you test it?**
- Tested on test cluster

**Potential risks**

**Is hotfix candidate?**
- No.

### CODE REVIEW

File: service/history/queues/reader_quotas.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		readerCallerToPriority[newReaderRequest(int32(readerId)).Caller] = readerId
+	}
+	defaultPriority := defaultReaderId + maxReaders - 1
+
+	return quotas.NewPriorityRateLimiter(
+		func(req quotas.Request) int {
+			if priority, ok := readerCallerToPriority[req.Caller]; ok {
+				return priority
+			}
+			return defaultPriority
```
Comment (yux0) [inline service/history/queues/reader_quotas.go:56]: any case the default reader id != 0?

File: service/history/queues/reader_quotas.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	maxReaders int,
+) quotas.RequestRateLimiter {
+	rateLimiters := make(map[int]quotas.RateLimiter, maxReaders)
+	readerCallerToPriority := make(map[string]int, maxReaders)
+	for readerId := defaultReaderId; readerId != defaultReaderId+maxReaders; readerId++ {
+		// use readerId as priority
+		rateLimiters[readerId] = quotas.NewDefaultOutgoingRateLimiter(rateFn)
+		readerCallerToPriority[newReaderRequest(int32(readerId)).Caller] = readerId
+	}
+	defaultPriority := defaultReaderId + maxReaders - 1
```
Comment (yux0) [inline service/history/queues/reader_quotas.go:49]: lowestPriority?

**Review by yux0 — APPROVED** [review]

File: service/history/queues/reader_quotas.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+func NewReaderPriorityRateLimiter(
+	rateFn quotas.RateFn,
+	maxReaders int,
+) quotas.RequestRateLimiter {
+	rateLimiters := make(map[int]quotas.RateLimiter, maxReaders)
+	readerCallerToPriority := make(map[string]int, maxReaders)
+	for readerId := defaultReaderId; readerId != defaultReaderId+maxReaders; readerId++ {
+		// use readerId as priority
+		rateLimiters[readerId] = quotas.NewDefaultOutgoingRateLimiter(rateFn)
+		readerCallerToPriority[newReaderRequest(int32(readerId)).Caller] = readerId
```
Comment (yiminc) [inline service/history/queues/reader_quotas.go:47]: Took me a while to understand why we need newReaderRequest(int32(readerId)).Caller here.
We could add comment here that each reader will use ReaderID as Caller when using this rate limiter.

File: service/history/queues/reader_quotas.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+func newReaderRequest(
+	readerID int32,
+) quotas.Request {
+	return quotas.NewRequest(
+		"",
+		readerRequestToken,
+		strconv.Itoa(int(readerID)),
+		"",
+		"",
```
Comment (yiminc) [inline service/history/queues/reader_quotas.go:84]: Maybe add comment why only requestToken and Caller matters here.

**Review by yiminc — APPROVED** [review]

File: service/history/queues/reader_quotas.go (hunk 1 above)
Comment (yycptt) [reply service/history/queues/reader_quotas.go]: No, I don't think so. It's a const value and should not be changed.

**Pull request merged**